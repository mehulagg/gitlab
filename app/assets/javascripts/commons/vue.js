import Vue from 'vue';
import GlFeatureFlagsPlugin from '~/vue_shared/gl_feature_flags_plugin';

if (process.env.NODE_ENV !== 'production') {
  Vue.config.productionTip = false;
}

import IssuableApp from '~/issue_show/components/app.vue';
import RelatedMergeRequests from '~/related_merge_requests/components/related_merge_requests.vue';
Vue.component('issuable-app', IssuableApp);
Vue.component('related-merge-requests', RelatedMergeRequests);

const app = new Vue({
  el: '[data-vue-shell]'
});

Vue.use(GlFeatureFlagsPlugin);
