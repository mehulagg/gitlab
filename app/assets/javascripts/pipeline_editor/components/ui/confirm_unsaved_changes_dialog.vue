<script>
export default {
  props: {
    hasUnsavedChanges: {
      type: Boolean,
      required: true,
    },
  },
  created() {
    window.addEventListener('beforeunload', this.confirmChanges);
  },
  destroyed() {
    window.removeEventListener('beforeunload', this.confirmChanges);
  },
  methods: {
    confirmChanges(e) {
      if (this.hasUnsavedChanges) {
        e.preventDefault();
        e.returnValue = '';
      }
    },
  },
  render: () => null,
};
</script>
