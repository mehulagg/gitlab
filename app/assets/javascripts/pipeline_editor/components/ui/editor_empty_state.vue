<script>
import { GlSprintf } from '@gitlab/ui';
import { s__ } from '~/locale';

export default {
  components: {
    GlSprintf,
  },
  i18n: {
    title: s__('PipelineEditor|Optimize your workflow with CI/CD Pipelines'),
    body: s__(
      'PipelineEditor|Create a new %{codeStart}.gitlab-ci.yml%{codeEnd} file at the root of the repository to get started.',
    ),
  },
  inject: {
    emptyStateIllustrationPath: {
      default: '',
    },
  },
};
</script>
<template>
  <div class="gl-display-flex gl-flex-direction-column gl-align-items-center">
    <img :src="emptyStateIllustrationPath" data-testid="empty-state-illustration" />
    <h1 class="gl-font-size-h1">{{ $options.i18n.title }}</h1>
    <p>
      <gl-sprintf :message="$options.i18n.body" class="gl-display-inline">
        <template #code="{ content }">
          <code>{{ content }}</code>
        </template>
      </gl-sprintf>
    </p>
  </div>
</template>
