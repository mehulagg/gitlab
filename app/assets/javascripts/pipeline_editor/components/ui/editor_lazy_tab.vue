<script>
import { GlTab } from '@gitlab/ui';

/**
 * Wrapper of <gl-tab> to lazily render this tab's content
 * when shown **without dismounting it after**.
 *
 * Once the tab is selected it is permanently set as "not-lazy"
 * so it's contents are not dismounted.
 *
 * Usage:
 *
 * Same as <gl-tab>, for example:
 *
 * <gl-tabs>
 *   <editor-lazy-tab title="Tab 1"> ... </editor-lazy-tab>
 *   <editor-lazy-tab title="Tab 2"> ... </editor-lazy-tab>
 * </gl-tabs>
 */
export default {
  components: {
    GlTab,
  },
  props: {
    lazy: {
      type: Boolean,
      required: false,
      default: true,
    },
  },
  data() {
    return {
      isLazy: this.lazy,
    };
  },
  methods: {
    onClick() {
      this.isLazy = false;
    },
  },
};
</script>
<template>
  <gl-tab :lazy="isLazy" v-bind="$attrs" v-on="$listeners" @click="onClick">
    <slot v-for="slot in Object.keys($slots)" :slot="slot" :name="slot"></slot>
  </gl-tab>
</template>
