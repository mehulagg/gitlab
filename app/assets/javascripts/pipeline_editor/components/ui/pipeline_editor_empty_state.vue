<script>
import { GlButton, GlSprintf } from '@gitlab/ui';
import { __ } from '~/locale';
import glFeatureFlagsMixin from '~/vue_shared/mixins/gl_feature_flags_mixin';

export default {
  components: {
    GlButton,
    GlSprintf,
  },
  i18n: {
    title: __('Optimize your workflow with CI/CD Pipelines'),
    body: __(
      'Create a new %{codeStart}.gitlab-ci.yml%{codeEnd} file at the root of the repository to get started.',
    ),
    btnText: __('Create new CI/CD pipeline'),
  },
  mixins: [glFeatureFlagsMixin()],
  inject: {
    emptyStateIllustrationPath: {
      default: '',
    },
  },
  computed: {
    showCTAButton() {
      return this.glFeatures.pipelineEditorEmptyStateAction;
    },
  },
  methods: {
    createEmptyConfigFile() {
      this.$emit('createEmptyConfigFile');
    },
  },
};
</script>
<template>
  <div class="gl-display-flex gl-flex-direction-column gl-align-items-center gl-mt-11">
    <img :src="emptyStateIllustrationPath" />
    <h1 class="gl-font-size-h1">{{ $options.i18n.title }}</h1>
    <p class="gl-mt-3">
      <gl-sprintf :message="$options.i18n.body">
        <template #code="{ content }">
          <code>{{ content }}</code>
        </template>
      </gl-sprintf>
    </p>
    <gl-button
      v-if="showCTAButton"
      variant="confirm"
      class="gl-mt-3"
      @click="createEmptyConfigFile"
    >
      {{ $options.i18n.btnText }}
    </gl-button>
  </div>
</template>
