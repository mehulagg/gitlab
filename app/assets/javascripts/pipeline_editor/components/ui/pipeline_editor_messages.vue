<script>
import { GlAlert } from '@gitlab/ui';
import { getParameterValues, removeParams } from '~/lib/utils/url_utility';
import { __, s__ } from '~/locale';
import {
  COMMIT_FAILURE,
  COMMIT_SUCCESS,
  DEFAULT_FAILURE,
  LOAD_FAILURE_UNKNOWN,
} from '../../constants';
import CodeSnippetAlert from '../code_snippet_alert/code_snippet_alert.vue';
import {
  CODE_SNIPPET_SOURCE_URL_PARAM,
  CODE_SNIPPET_SOURCES,
} from '../code_snippet_alert/constants';

export default {
  components: {
    GlAlert,
    CodeSnippetAlert,
  },
  errorTexts: {
    [COMMIT_FAILURE]: s__('Pipelines|The GitLab CI configuration could not be updated.'),
    [DEFAULT_FAILURE]: __('Something went wrong on our end.'),
    [LOAD_FAILURE_UNKNOWN]: s__('Pipelines|The CI configuration was not loaded, please try again.'),
  },
  successTexts: {
    [COMMIT_SUCCESS]: __('Your changes have been successfully committed.'),
  },
  props: {
    failureType: {
      type: String,
      required: false,
      default: null,
    },
    failureReasons: {
      type: Array,
      required: false,
      default: () => [],
    },
    successType: {
      type: String,
      required: false,
      default: null,
    },
  },
  data() {
    return {
      codeSnippetCopiedFrom: '',
      showAlert: false,
    };
  },
  computed: {
    alerts() {
      const messages = [];

      if (this.failureType) {
        messages.push(this.failure);
      }

      if (this.success) {
        messages.push(this.success);
      }
      return messages;
    },
    failure() {
      switch (this.failureType) {
        case LOAD_FAILURE_UNKNOWN:
          return {
            text: this.$options.errorTexts[LOAD_FAILURE_UNKNOWN],
            variant: 'danger',
          };
        case COMMIT_FAILURE:
          return {
            text: this.$options.errorTexts[COMMIT_FAILURE],
            variant: 'danger',
          };
        default:
          return {
            text: this.$options.errorTexts[DEFAULT_FAILURE],
            variant: 'danger',
          };
      }
    },
  },
  success() {
    switch (this.successType) {
      case COMMIT_SUCCESS:
        return {
          text: this.$options.successTexts[COMMIT_SUCCESS],
          variant: 'info',
        };
      default:
        return null;
    }
  },

  created() {
    this.parseCodeSnippetSourceParam();
  },
  methods: {
    dismissAlert() {
      this.showAlert = false;
    },
    dismissCodeSnippetAlert() {
      this.codeSnippetCopiedFrom = '';
    },
    parseCodeSnippetSourceParam() {
      const [codeSnippetCopiedFrom] = getParameterValues(CODE_SNIPPET_SOURCE_URL_PARAM);
      if (codeSnippetCopiedFrom && CODE_SNIPPET_SOURCES.includes(codeSnippetCopiedFrom)) {
        this.codeSnippetCopiedFrom = codeSnippetCopiedFrom;
        window.history.replaceState(
          {},
          document.title,
          removeParams([CODE_SNIPPET_SOURCE_URL_PARAM]),
        );
      }
    },
  },
};
</script>
<template>
  <div>
    <code-snippet-alert
      v-if="codeSnippetCopiedFrom"
      :source="codeSnippetCopiedFrom"
      class="gl-mb-5"
      @dismiss="dismissCodeSnippetAlert"
    />
    <gl-alert
      v-for="alert in alerts"
      :key="alert.text"
      :variant="alert.variant"
      class="gl-mb-5"
      @dismiss="dismissAlert"
    >
      {{ alert.text }}
      <ul v-if="failureReasons.length" class="gl-mb-0">
        <li v-for="reason in failureReasons" :key="reason">{{ reason }}</li>
      </ul>
    </gl-alert>
  </div>
</template>
