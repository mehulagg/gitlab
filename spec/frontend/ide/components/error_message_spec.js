import { mount } from '@vue/test-utils';
import { nextTick } from 'vue';
import ErrorMessage from '~/ide/components/error_message.vue';
import { GlLoadingIcon, GlAlert } from '@gitlab/ui';

const mockMessage = {
  text: 'some <strong>text</strong>',
  actionText: 'test action',
  actionPayload: 'testActionPayload',
};

const createComponent = function(messageOptions = {}, options = {}) {
  return mount(ErrorMessage, {
    propsData: {
      message: Object.assign({}, mockMessage, messageOptions),
    },
    ...options,
  });
};

describe('IDE error message component', () => {
  it('renders error message', () => {
    const wrapper = createComponent();
    expect(wrapper.text()).toContain('some text');
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('dismissal', () => {
    it('is not visible if there is an action available', () => {
      const actionMock = jest.fn();
      const wrapper = createComponent({ action: actionMock });
      expect(wrapper.find('button.gl-alert-dismiss').exists()).toBe(false);
    });

    it('will trigger the custom dismiss event', () => {
      const wrapper = createComponent();
      const setErrorMessageStub = jest.fn();
      wrapper.setMethods({ setErrorMessage: setErrorMessageStub });

      const alertComponent = wrapper.find(GlAlert);
      alertComponent.vm.$emit('dismiss');

      expect(setErrorMessageStub).toHaveBeenCalledWith(null);
    });
  });

  describe('messag actions', () => {
    it('will fire local action logic accordingly', () => {
      const actionMock = jest.fn().mockResolvedValue(true);
      const wrapper = createComponent({ action: actionMock });
      const actionButton = wrapper.find('button.gl-alert-action');

      actionButton.trigger('click');
      expect(actionMock).toHaveBeenCalledWith(mockMessage.actionPayload);
    });

    it('will correctly set loading state amd prevent further invocation', done => {
      let resolveAction;
      const actionMock = jest.fn().mockImplementation(
        () =>
          new Promise(resolve => {
            resolveAction = resolve;
          }),
      );
      const wrapper = createComponent({ action: actionMock });
      const actionButton = wrapper.find('button.gl-alert-action');
      const spinner = wrapper.find(GlLoadingIcon);

      expect(spinner.isVisible()).toBe(false);
      actionButton.trigger('click');
      expect(actionMock.mock.calls.length).toBe(1);

      return nextTick()
        .then(() => {
          expect(spinner.isVisible()).toBe(true);
          actionButton.trigger('click');
          // This click should not add another call to the message action
          expect(actionMock.mock.calls.length).toBe(1);
          resolveAction();
        })
        .then(nextTick())
        .then(() => {
          expect(spinner.isVisible()).toBe(false);
          done();
        });
    });
  });
});
