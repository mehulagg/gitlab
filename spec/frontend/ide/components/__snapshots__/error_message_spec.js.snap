// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IDE error message component renders error message 1`] = `
"<div role=\\"alert\\" class=\\"gl-alert gl-alert-danger\\"><svg class=\\"gl-icon s16 gl-alert-icon gl-alert-icon-no-title\\">
    <use href=\\"#error\\"></use>
  </svg> <button type=\\"button\\" aria-label=\\"Dismiss\\" class=\\"gl-alert-dismiss\\"><svg class=\\"gl-icon s16\\">
      <use href=\\"#close\\"></use>
    </svg></button>
  <!---->
  <div class=\\"gl-alert-body\\"><span>some <strong>text</strong></span>
    <glloadingicon-stub label=\\"Loading\\" size=\\"sm\\" color=\\"orange\\" inline=\\"true\\" style=\\"display: none;\\"></glloadingicon-stub>
  </div>
  <div class=\\"gl-alert-actions\\"><button type=\\"button\\" class=\\"btn gl-alert-action btn-danger btn-md new-gl-button\\">
      <!---->
      test action
    </button></div>
</div>"
`;
